### PR TITLE
Addon upgrade (partial implementation of #8107)

### PR DESCRIPTION
**Pull request by marekbiskup**
The minimal version. Probably not much more will go into v1.0, but we may want to resolve several questions.

limitations stated in kube-addon-update.sh.

What is handled:
1. create new addon services
2. delete removed addon services
3. create/delete/update addon replication controllers
   1. update if version is changed, e.g. if the name changes from my-rc-v1 to my-rc-v2
4. monitoring addon directory (it is checked every 5 minutes) - I don't know if this is necessary

What is missing:
1. node addons - we hope they will be handled by salt
2. update when only parameters changed, but the name didn't. (it is particularly important for services
3. restarting the linux service kube-addon after kube push (somehow it doesn't work)
4. tests (only manual tests were performed).
5. robustness (it is bash)
6. agreement on naming conventions for update - is versioning like "my-rc-v1" a good idea?
7. Agreement whether the monitoring of the add-on dir is necessary

Future extensions (probably not for)
1. rewrite in go
2. put into a pod
3. perform rolling update of replication controllers

### CODE REVIEW

**Comment by marekbiskup** [comment]
cc @satnam6502  @fgrzadkowski @zmerlynn @roberthbailey @davidopp @lavalamp

**Comment by marekbiskup** [comment]
Probably needs some polishing. I wanted to create the PR today, so that I can finish it tomorrow.

**Comment by zmerlynn** [comment]
Nice. I'll review it today.

**Comment by erictune** [comment]
Fixes #8811

**Comment by fgrzadkowski** [comment]
Ref #8107

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml [hunk 1]
```diff
@@ -1,18 +1,22 @@
 apiVersion: v1beta3
 kind: ReplicationController
 metadata: 
+  name: monitoring-heapster-v1
```
Comment (zmerlynn) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: meta-comment: if you're going to clean all these up, can you drop a README .. maybe in in cluster/addons ..  that describes the current contract of what's needed/wanted in an add-on?

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 2]
```diff
@@ -0,0 +1,390 @@
+#!/bin/bash
+
+# Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: nit: 2015

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+# Parameters
+# $1 path to add-ons
+
+
+# LIMITATIONS
+# 1. controllers are not updated unless their name is changed
+# 2. Services are treated as controllers, i.e. they will be udpated
+#    if service name is changed. the suffix after the last '-' is
+#    treated as verson. But we don't name services like this.
+# 3. Services will not be updated unless thir name is changed,
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: s/thir/their/

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+#    but for services we acually want updates without name change.
+# 4. Json files are not handled at all. Currently addons must be
+#    in yaml files
+# 5. exit code is prabably not always correct (I haven't checked
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: s/prabably/probably/

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+# echoes the string with paths to files containing addon for the given type
+# works only for yaml files (!) (ignores json files)
+function get-addons-from-disk() {
+    # prints to stdout, so log cannot be used
+    local -r addon_path=$1
+    local -r obj_type=$2
+
+    for filePath in $(find $addon_path -name \*.yaml); do
+        # WARNING: assumption that the topmost indentation is zero (I'm not sure yaml allows for topmost indentation)
+        cat "${filePath}" | grep "^kind: ${obj_type}" >/dev/null 2>/dev/null
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: I'd move this function to inline python, too.

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    for pid in $(jobs -p); do
+        wait ${pid} || (rv=1; log ERR "error in pid ${pid}")
+        log DB2 "pid ${pid} completed, current error code: $rv"
+    done
+    return $rv
+}
+
+
+function run-until-success() {
+    local -r command=$1;
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: unnecessary semicolons in this func

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+            echo $filePath
+        fi
+    done
+}
+
+# $1 file path
+function get-object-name-from-file() {
+    # prints to stdout, so log cannot be used
+    #WARNING: only yaml is supported
+    local -r name=$(cat $1 | python -c '''
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Split initialization of local and declaration to avoid errexit getting swallowed (bash gotcha). In this case, you can't use `local -r` either, because of this issue:

```
local name
name=$(blah)
```

If you do:
```
local -r name=$(foo)
```
the errexit can just get swallowed.

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+try:
+            import pipes,sys,yaml
+            y = yaml.load(sys.stdin)
+            print y["metadata"]["name"]
+except Exception, ex:
+            print "ERROR: %s" % ex
+        ''')
+    echo "${name}"
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Better yet, why are you round-tripping through a variable? You're doing..
```
local -r foo=$(bar)
echo "${foo}
```
but it could just be
```
bar
```
? (I don't think you're getting any whitspace squashing here?)

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+    # this will get the last field
+    echo "$input_string" | rev | cut -d "-" -f1 | rev
+}
+
+# returns the characters up to the last '-' (without it)
+# $1 input string
+function get-basename() {
+    # prints to stdout, so log cannot be used
+    local -r input_string=$1
+    local suffix=`get-suffix $input_string`
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: * Use `$(foo)` instead of `foo`.
* Split `local suffix` and `suffix=`

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+function get-addons-from-disk() {
+    # prints to stdout, so log cannot be used
+    local -r addon_path=$1
+    local -r obj_type=$2
+
+    for filePath in $(find $addon_path -name \*.yaml); do
+        # WARNING: assumption that the topmost indentation is zero (I'm not sure yaml allows for topmost indentation)
+        cat "${filePath}" | grep "^kind: ${obj_type}" >/dev/null 2>/dev/null 
+        if [[ $? -eq 0 ]]; then
+            echo $filePath
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: This is totally pedantic, but I'd prefer to see braces around variables when they're read. I'm not always the best adherent either, but it's a style thing.

So s/`$filePath`/`${filePath}`/ and many others.

File: test/e2e/monitoring.go [hunk 11]
```diff
@@ -74,17 +71,25 @@ var (
...
+	// update e.g. to heapster-monitoring-v2. By using a label query we can check for the
+	// situaiton when a heapster-monitoring-v1 and heapster-monitoring-v2 replication controller
+	// is running (which would be an error except during a rolling update).
+	for _, rcLabel := range rcLabels {
+		rcList, err := c.ReplicationControllers(api.NamespaceDefault).List(labels.Set{"k8s-app": rcLabel}.AsSelector())
+		if err != nil {
+			return nil, err
+		}
+		if len(rcList.Items) != 1 {
+			return nil, fmt.Errorf("expected to find one replicat for RC with label %s but got %d",
```
Comment (zmerlynn) [inline test/e2e/monitoring.go]: replica

**Comment by zmerlynn** [comment]
The approach looks good. If you can get me another version by your evening / my morning, I tend to start pretty early, we can probably go another round before you even sign off tomorrow - hopefully we can tie this up.

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 3 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 4 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 8 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 5 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 10 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: I agree. It doesn't matter so much which convention is better. But what matters is to have it consistent.
I'll make these changes.

ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 2 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 7 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 6 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 9 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: ok

File: test/e2e/monitoring.go (hunk 11 above)
Comment (marekbiskup) [inline test/e2e/monitoring.go]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+    local -r addon_path=$1
+    # be careful, reconcile-objects uses global variables
+    reconcile-objects $addon_path ReplicationController &
+
+    # Warning, we don't expect service names to be versioned, so
+    # the update for services should be turned off - just new services
+    # should be created and the old ones should be gone
+    # Also, if service description differs, the service should be recreated
+    # regardless of its name. This is not implemented in this version.
+    reconcile-objects $addon_path Service &
```
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: The currently implemented way of matching objects by basename is wrong for services. We have ```monitoring-influxdb``` and ```monitoring-grafana```. Currently they are matched to the same objects in different versions because the prefix ```monitoring-``` is the same.

I'll correct it to have the following semantics:
 * for rc: it will work as implemented
 * for services: services will match only if the name is exactly the same (no version after -)

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+  esac
+}
+
+
+
+# $1 addon path
+# $2 addon type (e.g. ReplicationController)
+# echoes the string with paths to files containing addon for the given type
+# works only for yaml files (!) (ignores json files)
+function get-addons-from-disk() {
```
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: I will skip all manifests that do not have "kubernetes.io/cluster-service: true". Otherwise it is impossible to perform upgrade by selecting objects with this label.

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml [hunk 14]
```diff
@@ -1,18 +1,22 @@
...
+  namespace: default
   labels: 
-    name: heapster
+    k8s-app: heapster
+    version: v1
     kubernetes.io/cluster-service: "true"
```
Comment (marekbiskup) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: -        kubernetes.io/cluster-service: "true"
+        kubernetes.io/cluster-service: true

Comment (marekbiskup) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: Actually I found that label values must be strings, so this has to stay as it is.

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 12 above)
Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: I didn't catch this, but why is the basename not everything before the version?

Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: I'm not sure I understand the question.

for rc, we have, for instance, monitoring-heapster-v1 (basename is monitoring-heapster), so update will be performed to monitoring-heapster-v2 (rolling update, in the future).
For services we can't do it because we have monitoring-influxdb (and no version suffix).

We can:
1) rename the services to follow this convention (But services are bound in a lot of places - a lot of changes)
or
2) declare that there will be no update for services - just create a new service and delete the old one if the name changes (this makes sense because there will be no rolling update anyway)

I decided to go for 2).

I think that the same can be done for rc. the version suffix is not necessary. But then we have to make sure rolling update works also if rc name doesn't change.

If I had more time I'd remove the version suffix from replication controllers too. (maybe it's still worth doing for 1.0?)

Comment (zmerlynn) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: I think we have time to remove the version suffix from RCs, and it's
probably still worth doing. Do you want to do it in this PR?

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml (hunk 1 above)
Comment (marekbiskup) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: ok

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 12 above)
Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Let's do it later. I won't be able to finish it today. Thursday and Friday are holidays. I can do it on Monday, if I'm permitted submit such changes.
versions can easily be removed also after 1.0 (just roling update will not work between "versioned" and "unversioned" replication controllers)

Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Ok, I have an idea how to do update without versions in the name:

1. match objects by name
2. compare metadata.labels.version
3. if changed, perform delete+create

*The only requirement for the user is to update the label "version" to a different value.*
For me it makes a lot of sense - much more than having version in the name. In the future we can extend it to any change inside the objects (if we find it useful).

I'll start preparing the RP (probably to be released on Monday)

Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: I found that addon-versioning is already used in kube-dns, which now is in version 2.

I think removing versions from controller names is a separate decision, so let's make it a separate PR too.

File: cluster/addons/README.md [hunk 15]
```diff
@@ -0,0 +1,45 @@
+# Cluster add-ons
+
+Cluster add-ons are Services and Replication Controllers (with pods) that are
+shipped with the kubernetes binaries and whose update policy is also consistent
+with the update of kubernetes cluster.
+
+On the clusterm the addons are kept in ```/usr/local/bin/kubectl``` on the master node, in yaml files
```
Comment (zmerlynn) [inline cluster/addons/README.md]: kept in `/etc/kubernetes/addons`, maybe?

**Comment by zmerlynn** [comment]
LGTM except a few nits.

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 12 above)
Comment (satnam6502) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Can you perform a rolling update without changing the name?

S

Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Good question. I don't know - I haven't actually tried, but I think I saw it in a doc somewhere (or maybe an issue at github).
I'll try that on Monday.

**Comment by marekbiskup** [comment]
@bgrant0607 clarified (in a personal email) that version string ``should be included`` in RC names

**Comment by marekbiskup** [comment]
Commits are sqashed. It's ready for merge.

**Comment by marekbiskup** [comment]
Sorry, I missed a few corrections. Still working on it.

**Comment by marekbiskup** [comment]
I'm going to include rotation of /var/log/kube-addon.log in my pull request. The check if addons update is pending is called every 10 minutes, so the file may get large after a few years...

**Comment by marekbiskup** [comment]
I found a way to restart a service by salt - I just have to create two salt states - one for stopping the service and another one for starting it. In the future it should be changed to monitoring files (I don't know why it doesn't work

**Comment by marekbiskup** [comment]
@zmerlynn, The change is (finally) finished on my side. Please have a look at it once again.

**Comment by zmerlynn** [comment]
LGTM! Let's try to get this in today.
